mpqp/execution/providers/azure.py

```diff
@@ -0,0 +1,195 @@
+    if job.job_type == JobType.SAMPLE:
+        assert job.measure is not None
+        job.status = JobStatus.RUNNING
+        job_sim = backend_sim.run(qiskit_circuit, shots=job.measure.shots)
+        result_sim = job_sim.result()
+        result = extract_result(result_sim, job, job.device)
+    else:
+        raise ValueError(f"Job type {job.job_type} not handled on Azure devices.")
+
+    job.status = JobStatus.DONE
+    return result
+
+
+@typechecked
+def extract_result(
+    result: "MicrosoftEstimatorResult | QiskitResult",
+    job: Optional[Job],
+    device: AZUREDevice,
+) -> Result:
+    from qiskit.result import Result as QiskitResult
+
+    if isinstance(result, QiskitResult):
+        if job is None:
+            job_data = result.data()
+            if "statevector" in job_data:
+                job_type = JobType.STATE_VECTOR
+                nb_qubits = int(math.log(len(result.get_statevector()), 2))
+                job = Job(job_type, QCircuit(nb_qubits), device)
+            elif "counts" in job_data:
+                job_type = JobType.SAMPLE
+                nb_qubits = len(list(result.get_counts())[0])
+                shots = result.results[0].shots
+                job = Job(
+                    job_type,
+                    QCircuit(nb_qubits),
+                    device,
+                    BasisMeasure(list(range(nb_qubits)), shots=shots),
+                )
+            else:
+                if len(result.data()) == 0:
+                    raise ValueError(
+                        "Result data is empty, cannot extract anything. Check "
+                        "if the associated job was successfully completed."
+                    )
+                else:
+                    raise ValueError(
+                        f"Data with keys {result.data().keys()} in result not handled."
+                    )
+
+        if job.job_type == JobType.STATE_VECTOR:
+            vector = np.array(result.get_statevector())
+            state_vector = StateVector(
+                vector,  # pyright: ignore[reportArgumentType]
+                job.circuit.nb_qubits,
+            )
+            return Result(job, state_vector, 0, 0)
+        elif job.job_type == JobType.SAMPLE:
+            assert job.measure is not None
+            job_data = result.data()
+            data = [
+                Sample(
+                    bin_str="".join(map(str, state)),
+                    nb_qubits=job.circuit.nb_qubits,
+                    count=int(count),
+                    probability=(
+                        job_data.get("probabilities").get(state)
+                        if "probabilities" in job_data
+                        else None
+                    ),
+                )
+                for (state, count) in job_data.get("counts").items()
+            ]
+            return Result(job, data, None, job.measure.shots)
+        else:
+            raise NotImplementedError(f"{job.job_type} not handled.")
+    else:
```

Comment:
if the result is a `MicrosoftEstimatorResult`, there is only one jobtype possible I suppose? (observable one ?)

If this assomption is correct, it looks like there is an error in the Result object you return line 135. You give first the job, and then 0 for the `data` argument of the constructor, and then result.data() for the `errors`. Is there an error here?

Reply:
`MicrosoftEstimatorResult` is not a normal result, nor a sample, nor an observable, nor a state vector. It just gives a message about the estimate. At the moment, we're putting the message in error -> str. But I think we'll have to create a different type of result.